# RETAIL PROJECT 

### IMPORTING LIBRARIES 

In [1]:
import numpy as np

In [2]:
import pandas as pd 

In [3]:
import matplotlib.pyplot as plt

In [4]:
import seaborn as sns 

In [5]:
import snowflake.connector 

In [7]:
import pandas_profiling 

C:\Users\Admin\anaconda3\Lib\site-packages\numba\core\decorators.py:262: NumbaDeprecationWarning: numba.generated_jit is deprecated. Please see the documentation at: https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-generated-jit for more information and advice on a suitable replacement.
  warnings.warn(msg, NumbaDeprecationWarning)
C:\Users\Admin\anaconda3\Lib\site-packages\visions\backends\shared\nan_handling.py:50: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
  @nb.jit


ImportError: cannot import name 'DataError' from 'pandas.core.base' (C:\Users\Admin\anaconda3\Lib\site-packages\pandas\core\base.py)

In [8]:
import getpass

In [10]:
%matplotlib inline 

connecting to snowflake 

In [13]:
conn = snowflake.connector.connect(
user =  'gaurav3sharma',
password = getpass.getpass('your snowflake password '), 
account = 'MYKIHKT-JJ59238',
database = 'RETAIL',
schema = 'PUBLIC',
wareshouse = 'COMPUTER_WH'
)

your snowflake password ········


CREATING CURSOR

In [14]:
cur = conn.cursor()

In [16]:
select_demographic_raw = 'select * from demographic '

In [17]:
select_campaign_desc_raw = 'SELECT * FROM CAMPAIGN_DESC_RAW '

In [18]:
select_campaign_raw = 'SELECT * FROM CAMPAIGN_RAW '

In [19]:
select_coupon_raw = 'SELECT * FROM COUPON_RAW'

In [20]:
select_coupon_redempt_raw = 'SELECT * FROM COUPON_REDEMPT_RAW '

In [21]:
select_product_raw = 'SELECT * FROM PRODUCT_RAW'

In [22]:
select_transaction_raw = 'SELECT * FROM TRANSACTION_RAW'

EXECUTING CURSOR ON ALL VARIABLE 

In [26]:
cur.execute(select_demographic_raw)
demographic_raw = cur.fetch_pandas_all()

In [27]:
demographic_raw

AGE_DESC MARITAL_STATUS_CODE INCOME_DESC HOMEOWNER_DESC  \
0         65+                   A      35-49K      Homeowner   
1       45-54                   A      50-74K      Homeowner   
2       25-34                   U      25-34K        Unknown   
3       25-34                   U      75-99K      Homeowner   
4       45-54                   B      50-74K      Homeowner   
...       ...                 ...         ...            ...   
2495    25-34                   U      15-24K      Homeowner   
2496    45-54                   U    125-149K      Homeowner   
2497    45-54                   U      50-74K      Homeowner   
2498    19-24                   A      35-49K         Renter   
2499    55-64                   U      35-49K        Unknown   

          HH_COMP_DESC HOUSEHOLD_SIZE_DESC KID_CATEGORY_DESC  HOUSEHOLD_KEY  
0     2 Adults No Kids                   2      None/Unknown              1  
1     2 Adults No Kids                   2      None/Unknown              7  
2        2 Adults Kids                   3                 1              8  
3        2 Adults Kids                   4                 2             13  
4        Single Female                   1      None/Unknown             16  
...                ...                 ...               ...            ...  
2495     2 Adults Kids                   4                 2            444  
2496      1 Adult Kids                  5+                3+            917  
2497           Unknown                   1      None/Unknown           1727  
2498  2 Adults No Kids                   2      None/Unknown             11  
2499  2 Adults No Kids                   2      None/Unknown            272  

[2500 rows x 8 columns]

In [28]:
cur.execute(select_campaign_desc_raw)
campaign_desc_raw = cur.fetch_pandas_all()

In [29]:
campaign_desc_raw

DESCRIPTION  CAMPAIGN  START_DAY  END_DAY
0        TypeB        24        659      719
1        TypeC        15        547      708
2        TypeB        25        659      691
3        TypeC        20        615      685
4        TypeB        23        646      684
5        TypeB        21        624      656
6        TypeB        22        624      656
7        TypeA        18        587      642
8        TypeB        19        603      635
9        TypeB        17        575      607
10       TypeC        14        531      596
11       TypeB        16        561      593
12       TypeA        13        504      551
13       TypeB        11        477      523
14       TypeB        12        477      509
15       TypeB        10        463      495
16       TypeB         9        435      467
17       TypeA         8        412      460
18       TypeB         7        398      432
19       TypeC         6        393      425
20       TypeC         3        356      412
21       TypeB         5        377      411
22       TypeB         4        372      404
23       TypeB         1        346      383
24       TypeB         2        351      383
25       TypeA        30        323      369
26       TypeB        29        281      334
27       TypeB        28        259      320
28       TypeC        27        237      300
29       TypeA        26        224      264
30       TypeB        24        659      719
31       TypeC        15        547      708
32       TypeB        25        659      691
33       TypeC        20        615      685
34       TypeB        23        646      684
35       TypeB        21        624      656
36       TypeB        22        624      656
37       TypeA        18        587      642
38       TypeB        19        603      635
39       TypeB        17        575      607
40       TypeC        14        531      596
41       TypeB        16        561      593
42       TypeA        13        504      551
43       TypeB        11        477      523
44       TypeB        12        477      509
45       TypeB        10        463      495
46       TypeB         9        435      467
47       TypeA         8        412      460
48       TypeB         7        398      432
49       TypeC         6        393      425
50       TypeC         3        356      412
51       TypeB         5        377      411
52       TypeB         4        372      404
53       TypeB         1        346      383
54       TypeB         2        351      383
55       TypeA        30        323      369
56       TypeB        29        281      334
57       TypeB        28        259      320
58       TypeC        27        237      300
59       TypeA        26        224      264

In [30]:
cur.execute(select_campaign_raw)
campaign_raw = cur.fetch_pandas_all()
campaign_raw

DESCRIPTION  HOUSEHOLD_KEY  CAMPAIGN
0          TypeA             17        26
1          TypeA             27        26
2          TypeA            212        26
3          TypeA            208        26
4          TypeA            192        26
...          ...            ...       ...
7203       TypeC           1803        15
7204       TypeC           1082        15
7205       TypeC            942        15
7206       TypeC            855        15
7207       TypeC            334        15

[7208 rows x 3 columns]

In [31]:
cur.execute(select_coupon_raw)
coupon_raw = cur.fetch_pandas_all()
coupon_raw

COUPON_UPC  PRODUCT_ID  CAMPAIGN
0       10000089061       27160         4
1       10000089064       27754         9
2       10000089073       28897        12
3       51800009050       28919        28
4       52100000076       28929        25
...             ...         ...       ...
124543  10000085478    13038802        18
124544  10000085479    13039368        18
124545  10000085478    13039553        18
124546  10000085479    13776332        18
124547  10000085477    13213390        18

[124548 rows x 3 columns]

In [32]:
cur.execute(select_coupon_redempt_raw)
coupon_redempt_raw = cur.fetch_pandas_all()
coupon_redempt_raw

HOUSEHOLD_KEY  DAY   COUPON_UPC  CAMPAIGN
0                 1  421  10000085364         8
1                 1  421  51700010076         8
2                 1  427  54200000033         8
3                 1  597  10000085476        18
4                 1  597  54200029176        18
...             ...  ...          ...       ...
2313           2496  592  54900050076        18
2314           2496  610  55100000013        18
2315           2500  449  53663200076         8
2316           2500  449  54300031076         8
2317           2500  624  55000013950        18

[2318 rows x 4 columns]

In [33]:
cur.execute(select_product_raw)
product_raw = cur.fetch_pandas_all()
product_raw

PRODUCT_ID  MANUFACTURER    DEPARTMENT     BRAND  \
0           25671             2       GROCERY  National   
1           26081             2  MISC. TRANS.  National   
2           26093            69        PASTRY   Private   
3           26190            69       GROCERY   Private   
4           26355            69       GROCERY   Private   
...           ...           ...           ...       ...   
92348    18293142          6384       DRUG GM  National   
92349    18293439          6393       DRUG GM  National   
92350    18293696          6406       DRUG GM  National   
92351    18294080          6442       DRUG GM  National   
92352    18316298           764       GROCERY  National   

                 COMMODITY_DESC           SUB_COMMODITY_DESC  \
0                      FRZN ICE          ICE - CRUSHED/CUBED   
1      NO COMMODITY DESCRIPTION  NO SUBCOMMODITY DESCRIPTION   
2                         BREAD         BREAD:ITALIAN/FRENCH   
3          FRUIT - SHELF STABLE                  APPLE SAUCE   
4                 COOKIES/CONES            SPECIALTY COOKIES   
...                         ...                          ...   
92348                 BOOKSTORE              PAPERBACK BOOKS   
92349                 BOOKSTORE            CHILDRENS LOW END   
92350                 BOOKSTORE        PAPERBACK BEST SELLER   
92351                 BOOKSTORE              PAPERBACK BOOKS   
92352              PAPER TOWELS       PAPER TOWELS & HOLDERS   

      CURR_SIZE_OF_PRODUCT  
0                    22 LB  
1                           
2                           
3                    50 OZ  
4                    14 OZ  
...                    ...  
92348                       
92349                       
92350                       
92351                       
92352                       

[92353 rows x 7 columns]

In [33]:
cur.execute(select_product_raw)
product_raw = cur.fetch_pandas_all()
product_raw

PRODUCT_ID  MANUFACTURER    DEPARTMENT     BRAND  \
0           25671             2       GROCERY  National   
1           26081             2  MISC. TRANS.  National   
2           26093            69        PASTRY   Private   
3           26190            69       GROCERY   Private   
4           26355            69       GROCERY   Private   
...           ...           ...           ...       ...   
92348    18293142          6384       DRUG GM  National   
92349    18293439          6393       DRUG GM  National   
92350    18293696          6406       DRUG GM  National   
92351    18294080          6442       DRUG GM  National   
92352    18316298           764       GROCERY  National   

                 COMMODITY_DESC           SUB_COMMODITY_DESC  \
0                      FRZN ICE          ICE - CRUSHED/CUBED   
1      NO COMMODITY DESCRIPTION  NO SUBCOMMODITY DESCRIPTION   
2                         BREAD         BREAD:ITALIAN/FRENCH   
3          FRUIT - SHELF STABLE                  APPLE SAUCE   
4                 COOKIES/CONES            SPECIALTY COOKIES   
...                         ...                          ...   
92348                 BOOKSTORE              PAPERBACK BOOKS   
92349                 BOOKSTORE            CHILDRENS LOW END   
92350                 BOOKSTORE        PAPERBACK BEST SELLER   
92351                 BOOKSTORE              PAPERBACK BOOKS   
92352              PAPER TOWELS       PAPER TOWELS & HOLDERS   

      CURR_SIZE_OF_PRODUCT  
0                    22 LB  
1                           
2                           
3                    50 OZ  
4                    14 OZ  
...                    ...  
92348                       
92349                       
92350                       
92351                       
92352                       

[92353 rows x 7 columns]

In [34]:
cur.execute(select_transaction_raw)
transaction_raw = cur.fetch_pandas_all()
transaction_raw

HOUSEHOLD_KEY    BASKET_ID  DAY  PRODUCT_ID  QUANTITY  SALES_VALUE  \
0                2375  26984851472    1     1004906         1         1.39   
1                2375  26984851472    1     1033142         1         0.82   
2                2375  26984851472    1     1036325         1         0.99   
3                2375  26984851472    1     1082185         1         1.21   
4                2375  26984851472    1     8160430         1         1.50   
...               ...          ...  ...         ...       ...          ...   
450411            916  29339646926  176      887342         1         1.50   
450412            916  29339646926  176      942778         1         2.99   
450413           2086  29339647754  176      868764         3         4.00   
450414           2086  29339647754  176     1093413         1         0.76   
450415           2086  29339647754  176     1126392         1         2.83   

        STORE_ID  RETAIL_DISC  TRANS_TIME  WEEK_NO  COUPON_DISC  \
0            364        -0.60        1631        1            0   
1            364         0.00        1631        1            0   
2            364        -0.30        1631        1            0   
3            364         0.00        1631        1            0   
4            364        -0.39        1631        1            0   
...          ...          ...         ...      ...          ...   
450411       436        -0.46        1732       26            0   
450412       436         0.00        1732       26            0   
450413       436        -0.77        2203       26            0   
450414       436         0.00        2203       26            0   
450415       436        -1.42        2203       26            0   

        COUPON_MATCH_DISC  
0                       0  
1                       0  
2                       0  
3                       0  
4                       0  
...                   ...  
450411                  0  
450412                  0  
450413                  0  
450414                  0  
450415                  0  

[450416 rows x 12 columns]

CLOSING THE CURSOR

In [35]:
cur.close()
conn.close()

In [40]:
demographic_raw.head(5) ##head(5) is like a limit 5 in sql

AGE_DESC MARITAL_STATUS_CODE INCOME_DESC HOMEOWNER_DESC      HH_COMP_DESC  \
0      65+                   A      35-49K      Homeowner  2 Adults No Kids   
1    45-54                   A      50-74K      Homeowner  2 Adults No Kids   
2    25-34                   U      25-34K        Unknown     2 Adults Kids   
3    25-34                   U      75-99K      Homeowner     2 Adults Kids   
4    45-54                   B      50-74K      Homeowner     Single Female   

  HOUSEHOLD_SIZE_DESC KID_CATEGORY_DESC  HOUSEHOLD_KEY  
0                   2      None/Unknown              1  
1                   2      None/Unknown              7  
2                   3                 1              8  
3                   4                 2             13  
4                   1      None/Unknown             16

In [38]:
product_raw.shape ## shape is used to check the no of rows and cols in the table 

(92353, 7)

checking null values 

In [42]:
demographic_raw.isnull().sum()

AGE_DESC               0
MARITAL_STATUS_CODE    0
INCOME_DESC            0
HOMEOWNER_DESC         0
HH_COMP_DESC           0
HOUSEHOLD_SIZE_DESC    0
KID_CATEGORY_DESC      0
HOUSEHOLD_KEY          0
dtype: int64

In [43]:
campaign_desc_raw.isnull().sum()

DESCRIPTION    0
CAMPAIGN       0
START_DAY      0
END_DAY        0
dtype: int64

In [46]:
campaign_raw.isnull().sum()

DESCRIPTION      0
HOUSEHOLD_KEY    0
CAMPAIGN         0
dtype: int64

In [47]:
coupon_raw.isnull().sum()

COUPON_UPC    0
PRODUCT_ID    0
CAMPAIGN      0
dtype: int64

In [49]:
coupon_redempt_raw.isnull().sum()

HOUSEHOLD_KEY    0
DAY              0
COUPON_UPC       0
CAMPAIGN         0
dtype: int64

In [50]:
product_raw.isnull().sum()

PRODUCT_ID              0
MANUFACTURER            0
DEPARTMENT              0
BRAND                   0
COMMODITY_DESC          0
SUB_COMMODITY_DESC      0
CURR_SIZE_OF_PRODUCT    0
dtype: int64

In [51]:
transaction_raw.isnull().sum()

HOUSEHOLD_KEY        0
BASKET_ID            0
DAY                  0
PRODUCT_ID           0
QUANTITY             0
SALES_VALUE          0
STORE_ID             0
RETAIL_DISC          0
TRANS_TIME           0
WEEK_NO              0
COUPON_DISC          0
COUPON_MATCH_DISC    0
dtype: int64

#### NO NULL VALUES WERE FOUND IN ALL THE TABLES 

USING DESCRIBE TO GET ALL THE VALUES OF 5 NUMBER SUMMARY

In [53]:
demographic_raw.describe()  ## total 2500 households participated 

HOUSEHOLD_KEY
count     2500.00000
mean      1250.50000
std        721.83216
min          1.00000
25%        625.75000
50%       1250.50000
75%       1875.25000
max       2500.00000

In [54]:
campaign_desc_raw.describe()  ##total 60 campaigns were launched for max 30 days 

CAMPAIGN   START_DAY     END_DAY
count  60.000000   60.000000   60.000000
mean   15.500000  463.866667  510.466667
std     8.728484  133.343887  136.558359
min     1.000000  224.000000  264.000000
25%     8.000000  356.000000  404.000000
50%    15.500000  470.000000  502.000000
75%    23.000000  587.000000  642.000000
max    30.000000  659.000000  719.000000

In [55]:
campaign_raw.describe()

HOUSEHOLD_KEY     CAMPAIGN
count    7208.000000  7208.000000
mean     1268.696726    15.659545
std       731.936391     6.949705
min         1.000000     1.000000
25%       644.000000     9.000000
50%      1263.000000    14.000000
75%      1924.000000    20.000000
max      2500.000000    30.000000

In [57]:
coupon_raw.describe()

COUPON_UPC    PRODUCT_ID       CAMPAIGN
count  1.245480e+05  1.245480e+05  124548.000000
mean   2.198225e+10  4.651277e+06      15.855935
std    1.961641e+10  4.843450e+06       6.015524
min    1.000009e+10  2.567100e+04       1.000000
25%    1.000009e+10  9.593030e+05      13.000000
50%    1.000009e+10  1.411451e+06      14.000000
75%    5.111108e+10  8.357538e+06      18.000000
max    5.998660e+10  1.814854e+07      30.000000

In [58]:
coupon_redempt_raw.describe()

HOUSEHOLD_KEY          DAY    COUPON_UPC     CAMPAIGN
count    2318.000000  2318.000000  2.318000e+03  2318.000000
mean     1302.816652   528.217429  4.123049e+10    15.551337
std       783.002545   100.360985  1.986068e+10     5.716636
min         1.000000   225.000000  1.000009e+10     1.000000
25%       588.000000   458.250000  1.000009e+10    13.000000
50%      1396.500000   532.000000  5.234003e+10    14.000000
75%      2004.000000   605.000000  5.430002e+10    18.000000
max      2500.000000   704.000000  5.897850e+10    30.000000

In [59]:
product_raw.describe()

PRODUCT_ID  MANUFACTURER
count  9.235300e+04   92353.00000
mean   5.328353e+06    1739.22833
std    5.359937e+06    1818.26957
min    2.567100e+04       1.00000
25%    9.706280e+05     328.00000
50%    1.621091e+06    1094.00000
75%    9.704770e+06    2264.00000
max    1.831630e+07    6477.00000

In [60]:
transaction_raw.describe()

HOUSEHOLD_KEY     BASKET_ID            DAY    PRODUCT_ID  \
count   450416.00000  4.504160e+05  450416.000000  4.504160e+05   
mean      1259.55822  2.840143e+10     112.470718  2.209464e+06   
std        728.10869  5.792110e+08      42.064258  2.726512e+06   
min          1.00000  2.698485e+10       1.000000  2.567100e+04   
25%        641.00000  2.798531e+10      84.000000  9.118780e+05   
50%       1234.00000  2.839168e+10     117.000000  1.016800e+06   
75%       1914.00000  2.886773e+10     147.000000  1.115360e+06   
max       2500.00000  2.933965e+10     176.000000  1.260019e+07   

            QUANTITY    SALES_VALUE       STORE_ID    RETAIL_DISC  \
count  450416.000000  450416.000000  450416.000000  450416.000000   
mean       86.522561       3.018436    2586.661648      -0.552774   
std      1058.788785       3.860427    7996.435988       1.186736   
min         0.000000       0.000000       1.000000     -75.000000   
25%         1.000000       1.250000     330.000000      -0.740000   
50%         1.000000       2.000000     369.000000      -0.090000   
75%         1.000000       3.390000     421.000000       0.000000   
max     51912.000000     400.000000   32124.000000       0.000000   

          TRANS_TIME        WEEK_NO    COUPON_DISC  COUPON_MATCH_DISC  
count  450416.000000  450416.000000  450416.000000      450416.000000  
mean     1583.582364      16.754487      -0.014860          -0.001920  
std       415.130098       6.018864       0.195378           0.044335  
min         0.000000       1.000000     -17.000000          -3.000000  
25%      1317.000000      13.000000       0.000000           0.000000  
50%      1630.000000      17.000000       0.000000           0.000000  
75%      1906.000000      22.000000       0.000000           0.000000  
max      2359.000000      26.000000       0.000000           0.000000

AVG QUANTITY IS  86 for this campaign . AVG SALE IS 3  ,  AVG RETAIL DISCOUNT IS  $0.5  ,AVG COUPON DISCOUNT IS 0.01 

## DATA MODIFICATION

STORING THE DATETIME IN A START_DATE VARIABLE 

In [61]:
START_DATE = pd.to_datetime ('2020-01-01')  ## setting the datetime format with pd.to_timedate 

In [62]:
START_DATE

Timestamp('2020-01-01 00:00:00')

In [63]:
transaction_raw.head(2)

HOUSEHOLD_KEY    BASKET_ID  DAY  PRODUCT_ID  QUANTITY  SALES_VALUE  \
0           2375  26984851472    1     1004906         1         1.39   
1           2375  26984851472    1     1033142         1         0.82   

   STORE_ID  RETAIL_DISC  TRANS_TIME  WEEK_NO  COUPON_DISC  COUPON_MATCH_DISC  
0       364         -0.6        1631        1            0                  0  
1       364          0.0        1631        1            0                  0

In [64]:
transaction_raw['DATE'] = START_DATE + pd.to_timedelta(transaction_raw['DAY'], unit = 'D')

In [68]:
transaction_raw.head(2) ## we can see that the new date column is added and date part is added day wise 

HOUSEHOLD_KEY    BASKET_ID  DAY  PRODUCT_ID  QUANTITY  SALES_VALUE  \
0           2375  26984851472    1     1004906         1         1.39   
1           2375  26984851472    1     1033142         1         0.82   

   STORE_ID  RETAIL_DISC  TRANS_TIME  WEEK_NO  COUPON_DISC  COUPON_MATCH_DISC  \
0       364         -0.6        1631        1            0                  0   
1       364          0.0        1631        1            0                  0   

        DATE  
0 2020-01-02  
1 2020-01-02

CREATING THE NEW START DATE AND END DATE COL IN CAMPAIGNDESC TABLE 

In [70]:
campaign_desc_raw['START_DATE'] = START_DATE + pd.to_timedelta(campaign_desc_raw['START_DAY'] , unit='D')

In [71]:
campaign_desc_raw.head(2)

DESCRIPTION  CAMPAIGN  START_DAY  END_DAY START_DATE
0       TypeB        24        659      719 2021-10-21
1       TypeC        15        547      708 2021-07-01

In [72]:
campaign_desc_raw['END_DATE'] = START_DATE + pd.to_timedelta(campaign_desc_raw['END_DAY'] , unit='D')

In [73]:
campaign_desc_raw.head(2)

DESCRIPTION  CAMPAIGN  START_DAY  END_DAY START_DATE   END_DATE
0       TypeB        24        659      719 2021-10-21 2021-12-20
1       TypeC        15        547      708 2021-07-01 2021-12-09

In [75]:
coupon_redempt_raw['DATE'] = START_DATE + pd.to_timedelta(coupon_redempt_raw['DAY'] , unit='D')

In [76]:
coupon_redempt_raw.head(2)

HOUSEHOLD_KEY  DAY   COUPON_UPC  CAMPAIGN       DATE
0              1  421  10000085364         8 2021-02-25
1              1  421  51700010076         8 2021-02-25

In [77]:
campaign_desc_raw['CAMPAIGN_DURATION'] = campaign_desc_raw['END_DATE'] - campaign_desc_raw['START_DATE'] 

In [80]:
campaign_desc_raw.head(2) ## new column campaign duration has been created

DESCRIPTION  CAMPAIGN  START_DAY  END_DAY START_DATE   END_DATE  \
0       TypeB        24        659      719 2021-10-21 2021-12-20   
1       TypeC        15        547      708 2021-07-01 2021-12-09   

  CAMPAIGN_DURATION  
0           60 days  
1          161 days

In [80]:
campaign_desc_raw.head(2) ## new column campaign duration has been created

DESCRIPTION  CAMPAIGN  START_DAY  END_DAY START_DATE   END_DATE  \
0       TypeB        24        659      719 2021-10-21 2021-12-20   
1       TypeC        15        547      708 2021-07-01 2021-12-09   

  CAMPAIGN_DURATION  
0           60 days  
1          161 days

In [81]:
demographic_raw.head(2)

AGE_DESC MARITAL_STATUS_CODE INCOME_DESC HOMEOWNER_DESC      HH_COMP_DESC  \
0      65+                   A      35-49K      Homeowner  2 Adults No Kids   
1    45-54                   A      50-74K      Homeowner  2 Adults No Kids   

  HOUSEHOLD_SIZE_DESC KID_CATEGORY_DESC  HOUSEHOLD_KEY  
0                   2      None/Unknown              1  
1                   2      None/Unknown              7

In [83]:
demographic_raw['AGE_DESC'].value_counts()  

AGE_DESC
45-54    901
35-44    594
25-34    445
65+      230
55-64    187
19-24    143
Name: count, dtype: int64

In [85]:
demographic_raw['HOMEOWNER_DESC'].value_counts()

HOMEOWNER_DESC
Homeowner          1573
Unknown             728
Renter              132
Probable Renter      34
Probable Owner       33
Name: count, dtype: int64

In [86]:
demographic_raw['HH_COMP_DESC'].value_counts()

HH_COMP_DESC
2 Adults No Kids    800
2 Adults Kids       584
Single Female       446
Single Male         292
Unknown             229
1 Adult Kids        149
Name: count, dtype: int64

In [87]:
demographic_raw['HOUSEHOLD_SIZE_DESC'].value_counts()

HOUSEHOLD_SIZE_DESC
2     1004
1      783
3      338
5+     206
4      169
Name: count, dtype: int64

In [88]:
demographic_raw['KID_CATEGORY_DESC'].value_counts()

KID_CATEGORY_DESC
None/Unknown    1737
1                357
3+               216
2                190
Name: count, dtype: int64

In [90]:
demographic_raw['INCOME_DESC'].value_counts()

INCOME_DESC
50-74K       588
35-49K       540
75-99K       305
15-24K       238
25-34K       237
Under 15K    193
125-149K     118
100-124K     109
150-174K      96
175-199K      33
250K+         29
200-249K      14
Name: count, dtype: int64

OBSERVATIONS: 901 BETWEEN AGE OF 45 TO 54 AND 143 BETWEEN 19-24 . 1573 ARE HOMEOWNEERS ADN AND 132 PEOPLE LIVE ON RENT 
588 MEMBERS EARN BETWEEN 50 TO 74k AND ONLY 14 MEMBERS EARN BETWEEN 200 -249K.800 MEMBERS HAVE NO KIDS WHERE AS 149 HAVE 1 ADULT KIDS . 216 MEMBERS HAVE 3+ KIDS AND 357 HAVE ONLY 1 KID . 1004 HOUSES HAVE 2 PEOPLE AND 169 WITH 4 MEMBERS 

FINDING THE START MONTH AND THE END MONTH WITH START YEAR AND END YEAR OF THE CAMPAIGN BY ADDING A NEW COLUMN

In [91]:
campaign_desc_raw['START_MONTH'] = campaign_desc_raw['START_DATE'].dt.strftime('%m')

In [92]:
campaign_desc_raw['END_MONTH'] = campaign_desc_raw['END_DATE'].dt.strftime('%m')

In [94]:

campaign_desc_raw['START_YEAR'] = campaign_desc_raw['START_DATE'].dt.strftime('%y')

In [95]:
campaign_desc_raw['END_YEAR'] = campaign_desc_raw['START_DATE'].dt.strftime('%y')

In [96]:
campaign_desc_raw.head(2)

DESCRIPTION  CAMPAIGN  START_DAY  END_DAY START_DATE   END_DATE  \
0       TypeB        24        659      719 2021-10-21 2021-12-20   
1       TypeC        15        547      708 2021-07-01 2021-12-09   

  CAMPAIGN_DURATION START_MONTH END_MONTH START_YEAR END_YEAR  
0           60 days          10        12         21       21  
1          161 days          07        12         21       21

In [105]:
campaign_desc_raw.groupby('DESCRIPTION').aggregate({'CAMPAIGN': 'count' , 'CAMPAIGN_DURATION':'mean'})

CAMPAIGN          CAMPAIGN_DURATION
DESCRIPTION                                     
TypeA              10           47 days 04:48:00
TypeB              38 37 days 15:09:28.421052631
TypeC              12           74 days 12:00:00

In [98]:
campaign_raw.head(2)

DESCRIPTION  HOUSEHOLD_KEY  CAMPAIGN
0       TypeA             17        26
1       TypeA             27        26

In [101]:
campaign_raw['DESCRIPTION'].value_counts()

DESCRIPTION
TypeA    3979
TypeB    2655
TypeC     574
Name: count, dtype: int64

In [103]:
campaign_raw['HOUSEHOLD_KEY'].nunique()

1584

OBSERVATION: 1584 HOUSES PARTICIPATED IN THE CAMPAING AND 916 DID NOT , TYPEA HAD 10 CAMPAIGNS FOR AVG OF 47 DAYS . TYPE B HAD 38 CAMPAIGNS FOR 37 DAYS AND TYPE C HAD MAX CAMPAIGNS LAUNCHED FOR 74 DAYS 

EDA FOR COUPON TABLE 

WE WILL CHECK HOW MANY CAMPAINS WERE LAUNCED  AND HOW MANY COUPONS WERE GIVEN IN THAT CAMPAIGN 

In [111]:
coupon_given = coupon_raw.groupby('CAMPAIGN').aggregate(total_products =('PRODUCT_ID','nunique'),total_coupon_given = ('COUPON_UPC','nunique'))

In [116]:
coupon_given.sort_values(by = 'total_products')

total_products  total_coupon_given
CAMPAIGN                                    
6                     18                   1
15                    55                   2
4                    200                  12
21                   242                  16
7                    259                   8
11                   278                  13
2                    301                  16
19                   375                  11
1                    383                  11
10                   393                  14
5                    443                  11
3                    506                  34
28                   535                  28
20                   609                  24
12                   620                  15
27                   632                  27
29                   689                  33
17                   713                  19
14                   714                  26
23                   767                  18
9                    825                  18
25                   826                  17
16                  1025                  13
24                  1887                   2
22                  1917                  21
30                  3205                 181
26                  3396                 181
8                  17178                 209
18                 35513                 209
13                 35616                 207

OBSERVATION : WE CAN SEE THAT CAMPAIGN 13, 18 AND 8 WERE THE CAMPAIGN WITH THE HIGHEST PRODUCTSS LAUNCHED AND THE MAXIMUM COUPONS WERE GIVEN 207 AND 209 IN BOTH 18 AND 8

####   JOINING COUPON_GIVEN  AND CAMPAIGN_DESC TABLE 

In [118]:
coupon_given = coupon_given.merge(right = campaign_desc_raw , on = 'CAMPAIGN' , how = 'left' )

In [120]:
coupon_given.head(2)

CAMPAIGN  total_products  total_coupon_given DESCRIPTION  START_DAY  \
0         1             383                  11       TypeB        346   
1         1             383                  11       TypeB        346   

   END_DAY START_DATE   END_DATE CAMPAIGN_DURATION START_MONTH END_MONTH  \
0      383 2020-12-12 2021-01-18           37 days          12        01   
1      383 2020-12-12 2021-01-18           37 days          12        01   

  START_YEAR END_YEAR  
0         20       20  
1         20       20

In [124]:
coupon_given.loc[:,('CAMPAIGN','total_products','total_coupon_given','DESCRIPTION','START_MONTH','END_MONTH','CAMPAIGN_DURATION')].sort_values(by ='total_products',ascending = False).head(10)

CAMPAIGN  total_products  total_coupon_given DESCRIPTION START_MONTH  \
25        13           35616                 207       TypeA          05   
24        13           35616                 207       TypeA          05   
35        18           35513                 209       TypeA          08   
34        18           35513                 209       TypeA          08   
15         8           17178                 209       TypeA          02   
14         8           17178                 209       TypeA          02   
51        26            3396                 181       TypeA          08   
50        26            3396                 181       TypeA          08   
59        30            3205                 181       TypeA          11   
58        30            3205                 181       TypeA          11   

   END_MONTH CAMPAIGN_DURATION  
25        07           47 days  
24        07           47 days  
35        10           55 days  
34        10           55 days  
15        04           48 days  
14        04           48 days  
51        09           40 days  
50        09           40 days  
59        01           46 days  
58        01           46 days

OBSERVATION : CAMPAIGN 13 LASTED FOR 47 DAYS BY STARTING MONTH OF MAY AND ENDED IN JULY WITH 35616 PRODUCTS BEING LAUNCHED .
CAMPAIGN 18 LASTED FOR 55 DAYS BY STARTING MONTH OF AUG AND ENDED IN OCT WITH 35513 PRODUCTS BEING LAUNCHED .
CAMPAIGN 8 LASTED FOR 48 DAYS BY STARTING MONTH OF FEB AND ENDED IN APRIL WITH 17718 PRODUCTS BEING LAUNCHED .

EDA FOR COUPON TABLE 

In [125]:
coupon_product = coupon_raw.merge(right = product_raw , on = 'PRODUCT_ID', how = 'left')

In [127]:
coupon_product.head(5)

COUPON_UPC  PRODUCT_ID  CAMPAIGN  MANUFACTURER DEPARTMENT     BRAND  \
0  10000089061       27160         4            69    GROCERY   Private   
1  10000089064       27754         9            69    GROCERY   Private   
2  10000089073       28897        12            69    GROCERY   Private   
3  51800009050       28919        28           236    GROCERY  National   
4  52100000076       28929        25           317    GROCERY  National   

             COMMODITY_DESC              SUB_COMMODITY_DESC  \
0            SHORTENING/OIL             VEGETABLE/SALAD OIL   
1    FRZN VEGETABLE/VEG DSH  FRZN BAGGED VEGETABLES - PLAIN   
2                      EGGS                  EGGS - X-LARGE   
3  REFRGRATD DOUGH PRODUCTS       REFRIGERATED COOKIES-CHUB   
4                    CHEESE                   STRING CHEESE   

  CURR_SIZE_OF_PRODUCT  
0                48 OZ  
1                16 OZ  
2           A D   1 DZ  
3              16.5 OZ  
4           AB    8 OZ

In [128]:
coupon_product['COMMODITY_DESC'].value_counts().head(10)

COMMODITY_DESC
HAIR CARE PRODUCTS        6261
MAKEUP AND TREATMENT      5585
FRZN MEAT/MEAT DINNERS    4893
FROZEN PIZZA              3658
BEEF                      3314
ICE CREAM/MILK/SHERBTS    3257
CHEESE                    3146
YOGURT                    2460
ORAL HYGIENE PRODUCTS     2437
LUNCHMEAT                 2425
Name: count, dtype: int64

EDA FOR COUPON_REDEMPED

In [130]:
coupon_redempt_raw['COUPON_UPC'].nunique()

556

556 coupons were redemed 

In [131]:
coupon_redemed = coupon_redempt_raw.groupby('CAMPAIGN').aggregate(total_coupon_redemend = ('COUPON_UPC','nunique'))

In [136]:
coupon_redemed.sort_values(by='total_coupon_redemend', ascending = False).head(10)

total_coupon_redemend
CAMPAIGN                       
13                          129
18                          121
8                           105
26                           53
30                           41
29                           17
23                           16
25                           15
20                           15
9                            14

OBSERVATION: campaign 13 had 129 coupons redeemed  , 18 had 121 and 8 had 105 

merging coupon reedemed and coupon given 

In [138]:
coupon_redemed = coupon_redemed.merge(right = coupon_given , on = 'CAMPAIGN',how = 'left')

In [140]:
coupon_redemed

CAMPAIGN  total_coupon_redemend  total_products  total_coupon_given  \
0          1                      1             383                  11   
1          1                      1             383                  11   
2          2                      4             301                  16   
3          2                      4             301                  16   
4          3                      2             506                  34   
5          3                      2             506                  34   
6          4                      6             200                  12   
7          4                      6             200                  12   
8          5                      5             443                  11   
9          5                      5             443                  11   
10         6                      1              18                   1   
11         6                      1              18                   1   
12         7                      5             259                   8   
13         7                      5             259                   8   
14         8                    105           17178                 209   
15         8                    105           17178                 209   
16         9                     14             825                  18   
17         9                     14             825                  18   
18        10                      7             393                  14   
19        10                      7             393                  14   
20        11                      6             278                  13   
21        11                      6             278                  13   
22        12                     11             620                  15   
23        12                     11             620                  15   
24        13                    129           35616                 207   
25        13                    129           35616                 207   
26        14                     14             714                  26   
27        14                     14             714                  26   
28        15                      2              55                   2   
29        15                      2              55                   2   
30        16                     10            1025                  13   
31        16                     10            1025                  13   
32        17                     13             713                  19   
33        17                     13             713                  19   
34        18                    121           35513                 209   
35        18                    121           35513                 209   
36        19                      9             375                  11   
37        19                      9             375                  11   
38        20                     15             609                  24   
39        20                     15             609                  24   
40        21                      5             242                  16   
41        21                      5             242                  16   
42        22                     12            1917                  21   
43        22                     12            1917                  21   
44        23                     16             767                  18   
45        23                     16             767                  18   
46        24                      2            1887                   2   
47        24                      2            1887                   2   
48        25                     15             826                  17   
49        25                     15             826                  17   
50        26                     53            3396                 181   
51        26                     53            3396                 181   
52        27                 

In [143]:
coupon_redemed['COUPON_REDEMPTION_RATE']=(coupon_redemed['total_coupon_redemend']/coupon_redemed['total_coupon_given'])*100

In [147]:
coupon_redemed.sort_values(by = 'COUPON_REDEMPTION_RATE', ascending = False ).head(2)

CAMPAIGN  total_coupon_redemend  total_products  total_coupon_given  \
47        24                      2            1887                   2   
46        24                      2            1887                   2   

   DESCRIPTION  START_DAY  END_DAY START_DATE   END_DATE CAMPAIGN_DURATION  \
47       TypeB        659      719 2021-10-21 2021-12-20           60 days   
46       TypeB        659      719 2021-10-21 2021-12-20           60 days   

   START_MONTH END_MONTH START_YEAR END_YEAR  COUPON_REDEMPTION_RATE  
47          10        12         21       21                   100.0  
46          10        12         21       21                   100.0

In [148]:
transaction_raw.head(2)

HOUSEHOLD_KEY    BASKET_ID  DAY  PRODUCT_ID  QUANTITY  SALES_VALUE  \
0           2375  26984851472    1     1004906         1         1.39   
1           2375  26984851472    1     1033142         1         0.82   

   STORE_ID  RETAIL_DISC  TRANS_TIME  WEEK_NO  COUPON_DISC  COUPON_MATCH_DISC  \
0       364         -0.6        1631        1            0                  0   
1       364          0.0        1631        1            0                  0   

        DATE  
0 2020-01-02  
1 2020-01-02

In [155]:
transaction_raw.drop(['DAY'],axis = 1 , inplace = True )

In [156]:
transaction_raw.head(2)

HOUSEHOLD_KEY    BASKET_ID  PRODUCT_ID  QUANTITY  SALES_VALUE  STORE_ID  \
0           2375  26984851472     1004906         1         1.39       364   
1           2375  26984851472     1033142         1         0.82       364   

   RETAIL_DISC  TRANS_TIME  WEEK_NO  COUPON_DISC  COUPON_MATCH_DISC       DATE  
0         -0.6        1631        1            0                  0 2020-01-02  
1          0.0        1631        1            0                  0 2020-01-02

In [160]:
campaign_desc_raw.drop(['START_DAY','END_DAY'],axis = 1 , inplace = True )

In [162]:
campaign_desc_raw.head(2)

DESCRIPTION  CAMPAIGN START_DATE   END_DATE CAMPAIGN_DURATION START_MONTH  \
0       TypeB        24 2021-10-21 2021-12-20           60 days          10   
1       TypeC        15 2021-07-01 2021-12-09          161 days          07   

  END_MONTH START_YEAR END_YEAR  
0        12         21       21  
1        12         21       21

In [166]:
coupon_redempt_raw.drop(['DAY'],axis = 1 , inplace = True)

In [167]:
coupon_redempt_raw.head(2)

HOUSEHOLD_KEY   COUPON_UPC  CAMPAIGN       DATE
0              1  10000085364         8 2021-02-25
1              1  51700010076         8 2021-02-25

CHANGING THE DATA TYPE OF NEW CREATED DATE COLUMNS TO OBJECT USING LAMBDA FUNCTION 

In [168]:
campaign_desc_raw['START_DATE'] =pd.to_datetime(campaign_desc_raw['START_DATE']).apply(lambda x :x.date())

In [169]:
campaign_desc_raw['END_DATE'] =pd.to_datetime(campaign_desc_raw['END_DATE']).apply(lambda x :x.date())

In [173]:
coupon_redempt_raw['DATE'] = pd.to_datetime(coupon_redempt_raw['DATE']).apply(lambda x :x.date())

In [174]:
transaction_raw['DATE'] = pd.to_datetime(transaction_raw['DATE']).apply(lambda x :x.date())

NOW LOADING THE NEW TABLE BACK TO  SNOWFLAKE 

IMPORTING LIBRARIES 

In [176]:
from sqlalchemy import create_engine

In [178]:
from sqlalchemy.engine import url 

In [179]:
import snowflake.connector as snowctx 

In [181]:
from snowflake.connector.pandas_tools import write_pandas

In [182]:
import pandas as pd

In [183]:
import getpass

ENTERING SNOWFLAKE DETAILS 

In [185]:
conn = snowflake.connector.connect(
user =  'gaurav3sharma',
password = getpass.getpass('your snowflake password '), 
account = 'MYKIHKT-JJ59238',
database = 'RETAIL',
schema = 'PUBLIC',
wareshouse = 'COMPUTER_WH'
)

your snowflake password ········


In [190]:
cur = conn.cursor()

In [191]:
cur.execute('''
CREATE OR REPLACE TABLE CAMPAIGN_DESC_NEW(
DESCRIPTION VARCHAR(30),
CAMPAIGN NUMBER(38,0),
START_DATE DATE , 
END_DATE DATE ,
CAMPAIGN_DURATION NUMBER(38,0),
START_MONTH NUMBER(38,0),
END_MONTH NUMBER(38,0),
START_YEAR NUMBER(38,0),
END_YEAR NUMBER(38,0))
''')

In [193]:
cur.execute('''
CREATE OR REPLACE TABLE COUPON_REDEMPT_NEW(
HOUSEHOLD_KEY NUMBER(38,0),
COUPON_UPC NUMBER(38,0),
CAMPAIGN NUMBER(38,0),
DATE DATE)
''')

In [198]:
cur.execute('''
CREATE OR REPLACE TABLE TRANSACTION_NEW(
HOUSEHOLD_KEY NUMBER(38,0), 
BASKET_ID NUMBER(38,0),
PRODUCT_ID NUMBER(38,0),
QUANTITY NUMBER(38,0),
SALES_VALUE FLOAT,
STORE_ID NUMBER(38,0),
RETAIL_DISC FLOAT , 
TRANS_TIME FLOAT ,
WEEK_NO NUMBER(38,0),
COUPON_DISC FLOAT , 
COUPON_MATCH_DISC FLOAT ,
DATE DATE )
''')

In [197]:
transaction_raw.head(1)

HOUSEHOLD_KEY    BASKET_ID  PRODUCT_ID  QUANTITY  SALES_VALUE  STORE_ID  \
0           2375  26984851472     1004906         1         1.39       364   

   RETAIL_DISC  TRANS_TIME  WEEK_NO  COUPON_DISC  COUPON_MATCH_DISC  \
0         -0.6        1631        1            0                  0   

         DATE  
0  2020-01-02

In [201]:
success, nchunks, nrows, _ = write_pandas(conn, campaign_desc_raw,'CAMPAIGN_DESC_NEW',quote_identifiers=False)
print(str(success)+','+str(nchunks)+','+str(nrows))

True,1,60


In [202]:
success, nchunks, nrows, _ = write_pandas(conn, coupon_redempt_raw,'COUPON_REDEMPT_NEW',quote_identifiers=False)
print(str(success)+','+str(nchunks)+','+str(nrows))


True,1,2318


In [204]:
success, nchunks, nrows, _ = write_pandas(conn,transaction_raw ,'TRANSACTION_NEW',quote_identifiers=False)
print(str(success)+','+str(nchunks)+','+str(nrows))

True,1,450416


In [205]:
cur.close()

True

In [206]:
conn.close()